core/src/main/scala/kafka/log/LogManager.scala

```diff
@@ -331,17 +418,22 @@ class LogManager(val logDirs: Array[File],
           log.truncateTo(truncateOffset)
           if (needToStopCleaner)
             cleaner.maybeTruncateCheckpoint(log.dir.getParentFile, topicPartition, log.activeSegment.baseOffset)
+        } catch {
+          case e: IOException =>
+            error(s"Failed to truncate log for $topicPartition due to IOException", e)
+            handleLogDirFailure(log.dir.getParent)
         } finally {
           if (needToStopCleaner)
             cleaner.resumeCleaning(topicPartition)
         }
       }
     }
-    checkpointRecoveryPointOffsets()
+    checkpointLogRecoveryOffsets()
   }
 
   /**
    *  Delete all data in a partition and start the log at the new offset
+   *
```

Comment:
Sure. I have updated the patch to catch IOException here.

Previously I had a long discussion with @becketqin and we think it is OK to handle IOException that currently doesn't trigger `halt()` in a followup patch. The reasons are recorded in this pull request. That is why I didn't catch IOException here.